media/server/service/source/main.cpp

```diff
@@ -47,13 +51,32 @@ int main(int argc, char *argv[])
 
     firebolt::rialto::server::gstInitalise(argc, argv);
 
-    auto appSessionServer =
-        firebolt::rialto::server::IApplicationSessionServerFactory::getFactory()->createApplicationSessionServer();
+    {
+        // Creation of this variable in a local scope ensures Rialto
+        // destructors are called before potentially calling gst_deinit()
+        // which would cause free memory reads
+        auto appSessionServer =
+            firebolt::rialto::server::IApplicationSessionServerFactory::getFactory()->createApplicationSessionServer();
+
+        if (!appSessionServer->init(argc, argv))
+        {
+            firebolt::rialto::server::gstDeinitalise();
+            return EXIT_FAILURE;
+        }
+
+        appSessionServer->startService();
+    }
 
-    if (!appSessionServer->init(argc, argv))
+#ifdef FREE_MEM_BEFORE_EXIT
+    if (getenv("RIALTO_FREE_MEM_BEFORE_EXIT"))
```

Comment:
Do we need both FREE_MEM_BEFORE_EXIT compiler flag and the env var? I would stay with the flag only.